plugins/com.gwtplugins.gdt.eclipse.core/src/com/google/gdt/eclipse/core/ReflectionUtilities.java

```diff
@@ -0,0 +1,16 @@
+/**
+ *
+ */
+package com.google.gdt.eclipse.core;
+
+/**
+ * @author luca
+ *
+ */
+public class ReflectionUtilities {
+
+  public static <T, R> R invoke(T obj, String method) throws ReflectiveOperationException {

```

Comment:
Since we always throw away the error (except when we wrap it with IllegalStateException, but we don't need to do that, since we already threw away the other error...), maybe just return Optional<R> instead, so that these can be chained?

Also, T doesn't serve a purpose here, probably remove it.

Finally, I'm not clear how often these methods are called, but if it is more than "once when you create the project" it might be worth caching the Method that succeeded so we can just call it for free next time, or even using a [MethodHandle](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/invoke/MethodHandle.html) and caching that instead? We could fail on the lookup that way (during static init of M2eCompat), and never have to try/catch when actually doing the work.

Reply:
The 'T' is what makes the whole thing work. If we cast a Record to an Object, the compiler complains, if it's a generic it does not. I wish I knew why. 
The correct solution is having a JDK 17 build, and a JDK 11 build, which is probaby not what we want.

MethodHandle and Optionals... it's not a bad idea.
About caching the handle at construction time, would you throw immediately or, say... use an Optional<MethodHandle> in order to fail later?

Reply:
i think you should not optimize code until you have a performance problem.

Reply:
The optimization here isn't for performance, but for complexity and brittleness - simpler code is easier for future maintainers to come back in 5+ years and understand what on earth we were thinking (which is very much a problem in other areas, see #453 for example). Avoiding duplication or checked-in compiled code makes it easier to only fix a bug in one place rather than have to search for other common patterns and clean them all up.

Pre-computing the method or methodhandle lets us fail fast, and, indicating right away that there is a problem in the constructor (or possibly slightly earlier) if the code we're runtime-linked against doesn't match either expected API. Assuming construction doesn't happen so early or in such a weird way that it otherwise breaks eclipse (that seems very unlikely to me, but worth testing). 

Making that change makes the Optional not really matter at all, since you already know the method exists (...as long as it doesn't throw, but we can unwrap such an exception and rethrow, at least according to the current contract of the two variants).

Reply:
@niloc132,
Quoting myself:

_As long as we use methods of this record, such as mavenProject(), everything is fine, but as soon as we try to use reflection, or even invoke getClass() the compiler complains because it says that we're referring java.lang.Record_

I'm really afraid that your proposal won't work. Not because it is wrong or whatever, but because we're using JDK 17 objects in JDK 11.

Reply:
Hmm. You're calling m.getClass() already though?

The `org.eclipse.m2e.core.project.configurator.ProjectConfigurationRequest` class is definitely being loaded and is initialized by the time you get to the assorted `configure()` methods, since you have an instance of it. 

Can you share the initialization error, or how to reproduce?

Reply:
Here the error, it happens at compile time, the class is definitely not loaded yet :)

```
Error:  Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:3.0.1:compile (default-compile) on project com.gwtplugins.gdt.eclipse.maven: Compilation failure: Compilation failure: 
Error:  /home/runner/work/gwt-eclipse-plugin/gwt-eclipse-plugin/plugins/com.gwtplugins.gdt.eclipse.maven/src/com/google/gdt/eclipse/maven/configurators/M2eCompat.java:[1] 
Error:  	package com.google.gdt.eclipse.maven.configurators;
Error:  	^
Error:  The type java.lang.Record cannot be resolved. It is indirectly referenced from required .class files
Error:  -> [Help 1]
```

https://github.com/gwt-plugins/gwt-eclipse-plugin/actions/runs/4199750062/jobs/7285005991#step:4:6450

Reply:
The point is that you can invoke getClass() on a ProjectConfigurationRequest, but *only* if you do it in a generic method.
As least... this is what I saw

Reply:
Hi @niloc132,
This is the commit which does not compile:

   https://github.com/gwt-plugins/gwt-eclipse-plugin/pull/451/commits/df2290106d34644a8afca4e3c09c27e00cc6e83b


This is the commit where I find the generic trick (sorry they've been split by mistake)

https://github.com/gwt-plugins/gwt-eclipse-plugin/pull/451/commits/f3bab4675b3ca33758c48d556632f98b61c87c8b
https://github.com/gwt-plugins/gwt-eclipse-plugin/pull/451/commits/e00e9cc6bdcac4913c2de3ca7f7e548a8a46f0c9

There's nothing smart in these commits, I'm just passing the ProjectConfigurationRequest  to a generic function which does not trigger the compilation error.

Reply:
So that's a compilation error, from referencing the class literal. As you're using the 2.x m2e jar on the classpath, but telling Java 17 that it should build to Java 11, it knows that this won't work - because it _doesn't_ know that when using Java 11 there will be a 1.x jar for m2e instead.

Instead, try `Class.forName("...ProjectConfigReq")` to get the Class instance up front. This probably can even be done in the core project, since it doesn't explicitly reference the class by name, and then you can test each method to see if it is there.